R/parity.R

```diff
@@ -1,66 +1,65 @@
+#'
 #' @importFrom foreach %do% %dopar% foreach
 #' @return numeric vector with the population parity for each column
-#' 
+#'
+#' @concept analyze
 #' @export
-redist.parity <- function(plans, total_pop, ncores = 1, 
-                          district_membership, population){
-  
-  if(!missing(population)){
-    .Deprecated(new = 'total_pop', old = 'population')
-    total_pop <- population
-  }
-  if(!missing(district_membership)){
-    .Deprecated(new = 'plans', old = 'district_membership')
-    plans <- district_membership
-  }
-  
-  if(!any(class(total_pop) %in% c('numeric', 'integer'))){
-    stop('Please provide "total_pop" as a numeric vector.')
-  }
-  
-  if(!any(class(plans) %in% c('numeric', 'matrix'))){
-    stop('Please provide "plans" as a matrix.')
-  }
-  
-  if(!is.matrix(plans)){
-    plans <- as.matrix(plans)
-  }
-  
-  if(length(total_pop) != nrow(plans)){
-    stop('Arguments "plans" and "total_pop" do not have same number of precincts.')
-  }
-  
-  # parallielze as in fastLink package to avoid Windows/unix issues
-  nc <- min(ncores, ncol(plans))
-  if (nc == 1){
-    `%oper%` <- `%do%`
-  } else { 
-    `%oper%` <- `%dopar%`
-    cl <- makeCluster(nc, setup_strategy = 'sequential')
-    registerDoParallel(cl)
-    on.exit(stopCluster(cl))
-  }
-  
-  out <- foreach(map = 1:ncol(plans)) %oper% {
-    # gets the district population using each column
-    distpop <- tapply(total_pop, plans[,map], sum)
-    # calculate the optimal pop
-    parpop <- sum(distpop) / length(distpop)
-    # calculate the pop parity
-    max(abs(distpop / parpop - 1))
-  }
-  
-  
-  return(unlist(out))
+redist.parity <- function(plans, total_pop, ncores=1, district_membership, population) {
+    if (!missing(population)) {
+        .Deprecated(new = 'total_pop', old = 'population')
+        total_pop <- population
+    }
+    if (!missing(district_membership)) {
+        .Deprecated(new = 'plans', old = 'district_membership')
+        plans <- district_membership
+    }
+
+    if (!any(class(total_pop) %in% c('numeric', 'integer'))) {
+        stop('Please provide "total_pop" as a numeric vector.')
+    }
+
+    if (!any(class(plans) %in% c('numeric', 'matrix'))) {
+        stop('Please provide "plans" as a matrix.')
+    }
+
+    if (!is.matrix(plans)) {
+        plans <- matrix(plans, ncol=1)
+    }
+
+    if (length(total_pop) != nrow(plans)) {
+        stop('Arguments "plans" and "total_pop" do not have same number of precincts.')
+    }
+
+    # parallielze as in fastLink package to avoid Windows/unix issues
+    N = ncol(plans)
+    nc <- min(ncores, floor(N/2))
+    if (nc == 1){
+        `%oper%` <- `%do%`
+    } else {
+        `%oper%` <- `%dopar%`
+        cl <- makeCluster(nc, setup_strategy = 'sequential')
+        registerDoParallel(cl)
+        on.exit(stopCluster(cl))
+    }
+
+    if (min(plans[,1]) == 0)
+        plans = plans + 1
+    n_distr = max(plans[,1])
+
+    chunks = split(1:N, rep(1:nc, each=ceiling(N/nc))[1:N])
```

Comment:
This is potentially breaking for maps with ncol(plans) < 2*nc. Is it better to copy the old code like this:
`  if(ncol(plans) < 2*nc){
out <- foreach(map = 1:ncol(plans)) %oper% {
    # gets the district population using each column
    distpop <- tapply(total_pop, plans[,map], sum)
    # calculate the optimal pop
    parpop <- sum(distpop) / length(distpop)
    # calculate the pop parity
    max(abs(distpop / parpop - 1))
  }
}`

Reply:
It's not the prettiest to have two versions, but it is a simple fix that gets directly at the potential problem.

Reply:
This still works for `N < 2*nc`. E.g. if `N=3` and `nc=8` you get
```
> chunks
$`1`
[1] 1

$`2`
[1] 2

$`3`
[1] 3
```

In testing, it's more efficient to break into `nc` chunks than to break into `N` chunks over `nc` cores.  Now, `nc` chunks of size 1 each is still probably faster to do sequentially, but if the user wants it...